src/task/future.rs

```diff
@@ -0,0 +1,405 @@
+// Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
+
+//! A [`Future`].
+
+use crate::{LocalSpawn, RemoteSpawn};
+
+use std::cell::UnsafeCell;
+use std::future::Future;
+use std::marker::PhantomData;
+use std::mem::ManuallyDrop;
+use std::pin::Pin;
+use std::sync::atomic::{AtomicU8, Ordering::SeqCst};
+use std::sync::Arc;
+use std::task::{Context, Poll, RawWaker, RawWakerVTable, Waker};
+use std::{fmt, mem};
+
+/// A [`Future`] task.
+pub struct Task<Remote, Extras> {
+    status: AtomicU8,
```

Comment:
How about defining an enum for the status?

Reply:
The status needs to support atomic operations. We cannot switch between enum variants atomically.